cpp/src/arrow/type.cc

```diff
@@ -1066,12 +1066,32 @@ struct FieldPathGetImpl {
     }
 
     int depth = 0;
-    const T* out;
+    const T* out = nullptr;
     for (int index : path->indices()) {
       if (children == nullptr) {
         return Status::NotImplemented("Get child data of non-struct array");
       }
 
+      if constexpr (std::is_same_v<T, std::shared_ptr<arrow::Field>>) {
+        // For lists, we don't care about the index, jump right into the list value type.
+        // The index here is used in the kernel to grab the specific list element.
+        // Children then are fields from list type, and out will be the children from
+        // that field, thus index 0.
+        if (out != nullptr && (*out)->type()->id() == Type::LIST) {
+          children = get_children(*out);
+          index = 0;
+        } else if (out == nullptr && children->size() == 1 && index >= 0) {
+          // WIP: Perhaps a dangerous assumption?
+          // For list types when previous was not a FieldPath, but
+          // a string name referencing a list type. Then the first iteration
+          // with index (referencing the list item), might look like it's
+          // out of bounds when actually we don't care about this index for
+          // type checking, only the kernel does; we need the list value type
+          // which is now the only child in the children vector.
```

Comment:
Can you show a concrete example of the case you are thinking of?

(also adding some tests for FieldPath::Get new capabilities (in C++) will probably be helpful to undestand the behaviour)

Reply:
Sure, this path happens with the `".a[1].c"` case, where the 'a' goes through string name, and the second '[1]' is through this FieldPath, but is also out of bounds from children (index 1) whereas `".a[0].c"` works b/c 0 is the index we want anyway from the list type. 

I'll work on adding the FieldPath::Get tests in C++. Thanks!

Reply:
But at this point we always have FieldPaths, no? (regardless whether they originally came from string names or not?)

Reply:
The first 'a', in our example, will be matched [here in the second overloaded member function](https://github.com/apache/arrow/blob/7ae4705c629655eb37b4200229d30f3a2af9e154/cpp/src/arrow/type.cc#L1407) then the second (`[1]`) goes to the first overloaded function taking `FieldPath`.